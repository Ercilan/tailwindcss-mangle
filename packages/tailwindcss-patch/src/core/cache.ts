import type { CacheOptions, InternalCacheOptions } from '../types'
import process from 'node:process'
import fs from 'fs-extra'
import path from 'pathe'
import { pkgName } from '../constants'
import logger from '../logger'

export function getCacheOptions(options?: CacheOptions | boolean) {
  let cache: InternalCacheOptions
  switch (typeof options) {
    case 'undefined': {
      cache = {
        enable: false,
      }
      break
    }
    case 'boolean': {
      cache = {
        enable: options,
      }
      break
    }
    case 'object': {
      cache = { ...options, enable: true }
      break
    }
  }
  return cache
}

export class CacheManager {
  public options: Required<CacheOptions> & { filename: string }

  constructor(options: CacheOptions = {}) {
    this.options = this.getOptions(options)
  }

  getOptions(options: CacheOptions = {}): Required<CacheOptions> & { filename: string } {
    const cwd = options.cwd ?? process.cwd()
    const dir = options.dir ?? path.resolve(cwd, 'node_modules/.cache', pkgName)
    const file = options.file ?? 'index.json'
    const filename = path.resolve(dir, file)
    return {
      cwd,
      dir,
      file,
      filename,
      strategy: 'merge',
    }
  }

  async write(data: Set<string>) {
    try {
      const { filename } = this.options
      await fs.outputJSON(filename, [...data])
      return filename
    }
    catch (error) {
      logger.error(error)
    }
  }

  async read() {
    const { filename } = this.options
    const isExisted = await fs.exists(filename)
    try {
      if (isExisted) {
        const data = await fs.readJSON(filename)
        return new Set<string>(data ?? [])
      }
    }
    catch {
      try {
        isExisted && await fs.remove(filename)
      }
      catch (error) {
        logger.error(error)
      }
    }
    return new Set<string>()
  }
}
