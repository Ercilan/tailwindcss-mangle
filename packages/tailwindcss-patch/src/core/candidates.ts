import type { SourceEntry } from '@tailwindcss/oxide'
import process from 'node:process'
import { defuOverrideArray } from '@tailwindcss-mangle/shared'

function importNode() {
  return import('@tailwindcss/node')
}

function importOxide() {
  return import('@tailwindcss/oxide')
}

export async function extractRawCandidatesWithPositions(
  content: string,
  extension: string = 'html',
): Promise<{ rawCandidate: string, start: number, end: number }[]> {
  const { Scanner } = await importOxide()
  const scanner = new Scanner({})

  const result = scanner.getCandidatesWithPositions({ content, extension })

  const candidates: { rawCandidate: string, start: number, end: number }[] = []
  for (const { candidate: rawCandidate, position: start } of result) {
    candidates.push({ rawCandidate, start, end: start + rawCandidate.length })
  }
  return candidates
}

export async function extractRawCandidates(
  sources?: SourceEntry[],
): Promise<string[]> {
  const { Scanner } = await importOxide()
  const scanner = new Scanner({
    sources,
  })

  const candidates = scanner.scan()

  return candidates
}

export interface ExtractValidCandidatesOption {
  sources?: SourceEntry[]
  base?: string
  css?: string
  cwd?: string
}

export async function extractValidCandidates(options?: ExtractValidCandidatesOption) {
  const defaultCwd = options?.cwd ?? process.cwd()
  const { sources, base, css } = defuOverrideArray<
    Required<ExtractValidCandidatesOption>,
    Partial<ExtractValidCandidatesOption>[]
  >(
    // @ts-ignore
    options,
    {
      css: '@import "tailwindcss";',
      base: defaultCwd,
      sources: [
        {
          base: defaultCwd,
          pattern: '**/*',
          negated: false,
        },
      ],
    },
  )

  const { __unstable__loadDesignSystem } = await importNode()
  const designSystem = await __unstable__loadDesignSystem(css, { base })

  const candidates = await extractRawCandidates(sources)
  const validCandidates = candidates.filter(
    rawCandidate => designSystem.parseCandidate(rawCandidate).length > 0,
  )
  return validCandidates
}
