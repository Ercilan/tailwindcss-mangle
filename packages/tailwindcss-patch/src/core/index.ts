export * from './patcher'
export * from './patches'
export * from './runtime'
export * from './cache'
