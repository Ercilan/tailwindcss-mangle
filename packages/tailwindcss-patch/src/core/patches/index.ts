export * from './exportContext'
export * from './supportCustomUnits'
