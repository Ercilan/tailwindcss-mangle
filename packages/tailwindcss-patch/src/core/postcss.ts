import path from 'node:path'
import process from 'node:process'
import { jiti } from '@/jiti'
import { defu, requireResolve } from '@/utils'
import { lilconfig } from 'lilconfig'
import postcss from 'postcss'

export interface ProcessTailwindcssOptions {
  cwd?: string
  config?: string
}

export async function processTailwindcss(options: ProcessTailwindcssOptions) {
  const { config: userConfig, cwd } = defu<ProcessTailwindcssOptions, ProcessTailwindcssOptions[]>(options, {
    cwd: process.cwd(),
  })
  let config = userConfig
  // 没有具体指定的话，就走下面的分支
  if (!(typeof config === 'string' && path.isAbsolute(config))) {
    const moduleName = 'tailwind'
    const result = await lilconfig('tailwindcss', {
      searchPlaces: [
        `${moduleName}.config.js`,
        `${moduleName}.config.cjs`,
        `${moduleName}.config.mjs`,
        `${moduleName}.config.ts`,
        `${moduleName}.config.cts`,
        `${moduleName}.config.mts`,
      ],
      loaders: {
        // 默认支持 js 和 cjs 2种格式
        '.js': jiti,
        '.cjs': jiti,
        '.mjs': jiti,
        '.ts': jiti,
        '.cts': jiti,
        '.mts': jiti,
      },
    }).search(cwd)
    if (!result) {
      throw new Error(`No TailwindCSS Config found in: ${cwd}`)
    }
    config = result.filepath
  }
  const id = requireResolve('tailwindcss', {
    basedir: cwd,
  })
  return await postcss([
    // eslint-disable-next-line ts/no-require-imports
    require(id)({
      config,
    }),
  ]).process('@tailwind base;@tailwind components;@tailwind utilities;', {
    from: undefined,
  })
}
