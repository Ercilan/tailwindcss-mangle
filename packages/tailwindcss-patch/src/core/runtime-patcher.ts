import path from 'node:path'
import fs from 'node:fs'
import process from 'node:process'
import { gte } from 'semver'
import type { PackageJson } from 'pkg-types'
import { defu } from 'defu'
import { inspectPostcssPlugin, inspectProcessTailwindFeaturesReturnContext } from './inspector'
import { inspectPostcssPlugin as inspectPostcssPluginCompat, inspectProcessTailwindFeaturesReturnContext as inspectProcessTailwindFeaturesReturnContextCompat } from './inspector-postcss7-compat'
import type { InternalPatchOptions, PatchOptions } from '@/types'
import { getDefaultPatchOptions } from '@/defaults'
import { ensureFileContent, requireResolve } from '@/utils'

export function getInstalledPkgJsonPath(options: PatchOptions = {}) {
  try {
    // const cwd = process.cwd()
    const tmpJsonPath = requireResolve(`tailwindcss/package.json`, {
      paths: options.paths,
      basedir: options.basedir,
    })

    return tmpJsonPath
    // https://github.com/sonofmagic/weapp-tailwindcss-webpack-plugin
    // only tailwindcss version > 3.0.0
  }
  catch (error) {
    if ((<Error & { code: string }>error).code === 'MODULE_NOT_FOUND') {
      console.warn('Can\'t find npm pkg: `tailwindcss`, Please ensure it has been installed!')
    }
  }
}

export function getPatchOptions(options: PatchOptions = {}) {
  return defu(
    options,
    {
      basedir: process.cwd(),
    },
    getDefaultPatchOptions(),
  ) as InternalPatchOptions
}

export function createPatch(opt: InternalPatchOptions) {
  return () => {
    try {
      const pkgJsonPath = getInstalledPkgJsonPath(opt)
      return internalPatch(pkgJsonPath, opt)
    }
    catch (error) {
      console.warn(`patch tailwindcss failed:${(<Error>error).message}`)
    }
  }
}

export function monkeyPatchForExposingContextV3(twDir: string, opt: InternalPatchOptions) {
  const processTailwindFeaturesFilePath = path.resolve(twDir, 'lib/processTailwindFeatures.js')

  const processTailwindFeaturesContent = ensureFileContent(processTailwindFeaturesFilePath)
  const result: { processTailwindFeatures?: string, plugin?: string } & Record<string, any> = {}
  if (processTailwindFeaturesContent) {
    const { code, hasPatched } = inspectProcessTailwindFeaturesReturnContext(processTailwindFeaturesContent)
    if (!hasPatched && opt.overwrite) {
      fs.writeFileSync(processTailwindFeaturesFilePath, code, {
        encoding: 'utf8',
      })
      console.log('patch tailwindcss processTailwindFeatures for return content successfully!')
    }
    result.processTailwindFeatures = code
  }

  const pluginFilePath = path.resolve(twDir, 'lib/plugin.js')
  const indexFilePath = path.resolve(twDir, 'lib/index.js')
  const pluginContent = ensureFileContent([pluginFilePath, indexFilePath])
  if (pluginContent) {
    const { code, hasPatched } = inspectPostcssPlugin(pluginContent)
    if (!hasPatched && opt.overwrite) {
      fs.writeFileSync(pluginFilePath, code, {
        encoding: 'utf8',
      })
      console.log('patch tailwindcss for expose runtime content successfully!')
    }
    result.plugin = code
  }

  opt.custom && typeof opt.custom === 'function' && opt.custom(twDir, result)
  return result
}

export function monkeyPatchForExposingContextV2(twDir: string, opt: InternalPatchOptions) {
  const processTailwindFeaturesFilePath = path.resolve(twDir, 'lib/jit/processTailwindFeatures.js')

  const processTailwindFeaturesContent = ensureFileContent(processTailwindFeaturesFilePath)
  const result: { processTailwindFeatures?: string, plugin?: string } & Record<string, any> = {}
  if (processTailwindFeaturesContent) {
    const { code, hasPatched } = inspectProcessTailwindFeaturesReturnContextCompat(processTailwindFeaturesContent)
    if (!hasPatched && opt.overwrite) {
      fs.writeFileSync(processTailwindFeaturesFilePath, code, {
        encoding: 'utf8',
      })
      console.log('patch tailwindcss processTailwindFeatures for return content successfully!')
    }
    result.processTailwindFeatures = code
  }

  const indexFilePath = path.resolve(twDir, 'lib/jit/index.js')
  const pluginContent = ensureFileContent([indexFilePath])
  if (pluginContent) {
    const { code, hasPatched } = inspectPostcssPluginCompat(pluginContent)
    if (!hasPatched && opt.overwrite) {
      fs.writeFileSync(indexFilePath, code, {
        encoding: 'utf8',
      })
      console.log('patch tailwindcss for expose runtime content successfully!')
    }
    result.plugin = code
  }

  opt.custom && typeof opt.custom === 'function' && opt.custom(twDir, result)
  return result
}

export function internalPatch(pkgJsonPath: string | undefined, options: InternalPatchOptions): any | undefined {
  if (pkgJsonPath) {
    // eslint-disable-next-line ts/no-var-requires, ts/no-require-imports
    const pkgJson = require(pkgJsonPath) as PackageJson
    const twDir = path.dirname(pkgJsonPath)
    if (gte(pkgJson.version!, '3.0.0')) {
      options.version = pkgJson.version
      const result = monkeyPatchForExposingContextV3(twDir, options)
      return result
    }
    else if (gte(pkgJson.version!, '2.0.0')) {
      options.version = pkgJson.version
      const result = monkeyPatchForExposingContextV2(twDir, options)
      return result
    }
    // no sth
  }
}
