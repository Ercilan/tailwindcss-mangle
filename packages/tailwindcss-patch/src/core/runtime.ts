import path from 'node:path'
import { gte } from 'semver'
import type { PackageJson } from 'pkg-types'
import { monkeyPatchForExposingContextV2, monkeyPatchForExposingContextV3, monkeyPatchForSupportingCustomUnit } from './patches'

import type { InternalPatchOptions } from '@/types'

export function internalPatch(pkgJsonPath: string | undefined, options: InternalPatchOptions) {
  if (pkgJsonPath) {
    // eslint-disable-next-line ts/no-var-requires, ts/no-require-imports
    const pkgJson = require(pkgJsonPath) as PackageJson
    const twDir = path.dirname(pkgJsonPath)
    options.version = pkgJson.version
    if (gte(pkgJson.version!, '3.0.0')) {
      options.version = pkgJson.version

      if (options.applyPatches?.extendLengthUnits) {
        try {
          monkeyPatchForSupportingCustomUnit(twDir, {
            overwrite: options.overwrite,
          })
        }
        catch {

        }
      }

      if (options.applyPatches?.exportContext) {
        return monkeyPatchForExposingContextV3(twDir, options)
      }
    }
    else if (gte(pkgJson.version!, '2.0.0')) {
      if (options.applyPatches?.exportContext) {
        return monkeyPatchForExposingContextV2(twDir, options)
      }
    }
    // no sth
  }
}
