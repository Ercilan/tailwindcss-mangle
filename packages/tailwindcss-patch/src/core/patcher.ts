import type { PatchUserConfig } from '@tailwindcss-mangle/config'
import type { InternalCacheOptions, InternalPatchOptions, PackageInfo, TailwindcssClassCache, TailwindcssPatcherOptions, TailwindcssRuntimeContext } from '../types'
import { createRequire } from 'node:module'
import process from 'node:process'
import fs from 'fs-extra'
import { getPackageInfoSync } from 'local-pkg'
import path from 'pathe'
import { getPatchOptions } from '../defaults'
import logger from '../logger'
import { isObject } from '../utils'
import { CacheManager, getCacheOptions } from './cache'
import { extractValidCandidates } from './candidates'
import { processTailwindcss } from './postcss'
import { internalPatch } from './runtime'

const require = createRequire(import.meta.url)
export class TailwindcssPatcher {
  public rawOptions: TailwindcssPatcherOptions
  public cacheOptions: InternalCacheOptions
  public patchOptions: InternalPatchOptions
  public patch: () => void
  public cacheManager: CacheManager
  public packageInfo: PackageInfo
  public majorVersion?: number

  constructor(options: TailwindcssPatcherOptions = {}) {
    this.rawOptions = options
    this.cacheOptions = getCacheOptions(options.cache)
    this.patchOptions = getPatchOptions(options.patch)

    this.cacheManager = new CacheManager(this.cacheOptions)

    const packageInfo = getPackageInfoSync('tailwindcss')

    if (!packageInfo) {
      throw new Error('tailwindcss not found')
    }

    if (packageInfo.version) {
      this.majorVersion = Number.parseInt(packageInfo.version[0])
    }
    this.packageInfo = packageInfo
    this.patch = () => {
      try {
        return internalPatch(this.packageInfo?.packageJsonPath, this.patchOptions)
      }
      catch (error) {
        logger.error(`patch tailwindcss failed: ${(<Error>error).message}`)
      }
    }
  }

  setCache(set: Set<string>) {
    if (this.cacheOptions.enable) {
      return this.cacheManager.write(set)
    }
  }

  getCache() {
    return this.cacheManager.read()
  }

  getContexts(): TailwindcssRuntimeContext[] {
    if (this.packageInfo) {
      const distPath = path.join(this.packageInfo.rootPath, 'lib')
      let injectFilePath: string | undefined
      if (this.majorVersion === 2) {
        injectFilePath = path.join(distPath, 'jit/index.js')
      }
      else if (this.majorVersion === 3) {
        injectFilePath = path.join(distPath, 'plugin.js')
        if (!fs.existsSync(injectFilePath)) {
          injectFilePath = path.join(distPath, 'index.js')
        }
      }
      if (injectFilePath) {
        const mo = require(injectFilePath)
        if (mo.contextRef) {
          return mo.contextRef.value
        }
      }
    }

    return []
  }

  getClassCaches(): TailwindcssClassCache[] {
    const contexts = this.getContexts()
    return contexts.filter(x => isObject(x)).map(x => x.classCache)
  }

  async getClassCacheSet(options?: PatchUserConfig): Promise<Set<string>> {
    const classSet = new Set<string>()
    const { output, tailwindcss } = options ?? {}
    if (this.majorVersion === 4) {
      const { v4 } = tailwindcss ?? {}
      if (Array.isArray(v4?.cssEntries)) {
        const results = (
          await Promise.all(
            v4
              .cssEntries
              .map(async (x) => {
                if (await fs.exists(x)) {
                  const css = await fs.readFile(x, 'utf8')
                  return css
                }
                return false
              }),
          )
        ).filter(x => x) as string[]
        for (const css of results) {
          const candidates = await extractValidCandidates({
            base: v4?.base,
            css,
            sources: v4?.sources?.map((x) => {
              return {
                base: x.base ?? v4?.base ?? process.cwd(),
                pattern: x.pattern,
              }
            }),
          })
          for (const candidate of candidates) {
            classSet.add(candidate)
          }
        }
      }
      else {
        const candidates = await extractValidCandidates({
          base: v4?.base,
          css: v4?.css,
          sources: v4?.sources?.map((x) => {
            return {
              base: x.base ?? v4?.base ?? process.cwd(),
              pattern: x.pattern,
            }
          }),
        })
        for (const candidate of candidates) {
          classSet.add(candidate)
        }
      }
    }
    else {
      const classCaches = this.getClassCaches()

      for (const classCacheMap of classCaches) {
        const keys = classCacheMap.keys()
        for (const key of keys) {
          const v = key.toString()
          if (output?.removeUniversalSelector && v === '*') {
            continue
          }
          classSet.add(v)
        }
      }
    }

    return classSet
  }

  /**
   * @description 在多个 tailwindcss 上下文时，这个方法将被执行多次，所以策略上应该使用 append
   */
  async getClassSet(options?: PatchUserConfig) {
    const { output, tailwindcss } = options ?? {}
    const cacheStrategy = this.cacheOptions.strategy ?? 'merge'
    const set = await this.getClassCacheSet({
      output,
      tailwindcss,
    })
    if (cacheStrategy === 'overwrite') {
      set.size > 0 && this.setCache(set)
    }
    else if (cacheStrategy === 'merge') {
      const cacheSet = this.getCache()
      if (cacheSet) {
        for (const x of cacheSet) {
          set.add(x)
        }
      }
      this.setCache(set)
    }

    return set
  }

  async extract(options?: PatchUserConfig) {
    const { output, tailwindcss } = options ?? {}
    if (output && tailwindcss) {
      const { filename, loose } = output

      if (this.majorVersion === 3) {
        await processTailwindcss({
          ...tailwindcss,
          majorVersion: this.majorVersion,
        })
      }

      const set = await this.getClassSet({
        output,
        tailwindcss,
      })
      if (filename) {
        const classList = [...set]
        await fs.outputJSON(filename, classList, {
          spaces: loose ? 2 : undefined,
        })
        return filename
      }
    }
  }

  extractValidCandidates = extractValidCandidates
}
