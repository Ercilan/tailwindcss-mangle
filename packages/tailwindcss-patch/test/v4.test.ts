import { extractRawCandidates } from '@/core/candidates'
import { TailwindcssPatcher } from '@/index'
import fs from 'fs-extra'
import path from 'pathe'
// import { __unstable__loadDesignSystem } from 'tailwindcss'

describe('v4', () => {
  it('tailwindcssPatcher case 0', async () => {
    const patcher = new TailwindcssPatcher()

    const candidates = await patcher.extractValidCandidates({
      base: import.meta.dirname,
      css: await fs.readFile(path.resolve(import.meta.dirname, './fixtures/v4/index.css'), 'utf8'),
      sources: [
        {
          base: import.meta.dirname,
          pattern: path.resolve(import.meta.dirname, './fixtures/v4/index.html'),
        },
      ],
    })
    expect(candidates).toMatchSnapshot()
  })

  it('tailwindcssPatcher case 1', async () => {
    const patcher = new TailwindcssPatcher()

    const candidates = await patcher.extractValidCandidates({
      base: import.meta.dirname,
      css: await fs.readFile(path.resolve(import.meta.dirname, './fixtures/v4/index.css'), 'utf8'),
      sources: [
        {
          base: import.meta.dirname,
          pattern: path.resolve(import.meta.dirname, './fixtures/v4/**/*.html'),
        },
      ],
    })
    expect(candidates).toMatchSnapshot()
  })

  it('tailwindcssPatcher getClassSet case 2', async () => {
    const patcher = new TailwindcssPatcher({
      patch: {
        tailwindcss: {
          v4: {
            base: import.meta.dirname,
            css: await fs.readFile(path.resolve(import.meta.dirname, './fixtures/v4/index.css'), 'utf8'),
            sources: [
              {
                base: import.meta.dirname,
                pattern: path.resolve(import.meta.dirname, './fixtures/v4/**/*.html'),
              },
            ],
          },
        },
      },
    })
    patcher.majorVersion = 4
    const candidates = await patcher.getClassSet()
    expect(candidates).toMatchSnapshot()
  })

  it('extractRawCandidates case 0', async () => {
    const candidates = await extractRawCandidates(
      [
        {
          base: import.meta.dirname,
          pattern: path.resolve(import.meta.dirname, './fixtures/v4/index.html'),
        },
      ],
    )
    expect(candidates).toMatchSnapshot()
  })

  // {
  //   base: import.meta.dirname,
  //   css: await fs.readFile(path.resolve(import.meta.dirname, './fixtures/v4/index.css'), 'utf8'),
  //   content: path.resolve(import.meta.dirname, './fixtures/v4/index.html'),
  // }
})
