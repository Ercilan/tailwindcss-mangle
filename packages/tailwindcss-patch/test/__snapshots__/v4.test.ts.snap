// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`v4 > extractRawCandidates case 0 1`] = `
[
  "!DOCTYPE",
  "bg-[#123456]",
  "body",
  "charset",
  "class",
  "content",
  "device-width",
  "div",
  "en",
  "head",
  "html",
  "initial-scale",
  "lang",
  "meta",
  "name",
  "text-(--x)",
  "text-xs",
  "title",
  "viewport",
  "width",
]
`;

exports[`v4 > tailwindcssPatcher casae 1 1`] = `
[
  "bg-[#123456]",
  "bg-gradient-to-b",
  "content",
  "from-[#2f73f1]",
  "h-[30px]",
  "text-(--x)",
  "text-xs",
  "to-[#4bcefd]",
  "w-[323px]",
]
`;

exports[`v4 > tailwindcssPatcher case 0 1`] = `
[
  "bg-[#123456]",
  "content",
  "text-(--x)",
  "text-xs",
]
`;
